### PR TITLE
Unit tests for lenient_lowercase function

### PR DESCRIPTION
**Pull request by Andersson007**
##### SUMMARY
Unit tests for lenient_lowercase function from lib/ansible/module_utils/common/text/formatters.py

##### ISSUE TYPE

- Bugfix Pull Request

### CODE REVIEW

**Comment by Andersson007** [comment]
@samdoran could you please look at this

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+def test_lenient_lowercase():
+    """Test of lenient_lowercase function"""
+    output_list = lenient_lowercase(INPUT_LIST)
+    for out_elem, exp_elem in zip(output_list, LOWERED_LIST):
+        assert out_elem == exp_elem
+
+
+if __name__ == '__main__':
+    test_lenient_lowercase()
```
Comment (webknjaz) [inline test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]: This is usually not needed

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    u'café',
+    b'hello',
+    1,
+    {1: 'Dict'},
+    True,
+    [1],
+]
+
+
+def test_lenient_lowercase():
```
Comment (webknjaz) [inline test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]: Please use `pytest.mark.parametrize()` instead of a loop

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+from ansible.module_utils.common.text.formatters import lenient_lowercase
+
+
+INPUT_LIST = [
+    u'HELLO',
+    u'Ёлка',
+    u'cafÉ',
```
Comment (webknjaz) [inline test/units/module_utils/common/text/formatters/test_lenient_lowercase.py:18]: it'd be nice to get a hieroglyphic test (Asian langs)

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    b'hello',
+    1,
+    {1: 'Dict'},
+    True,
+    [1],
+]
+
+
+def test_lenient_lowercase():
+    """Test of lenient_lowercase function"""
```
Comment (webknjaz) [inline test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]: As per PEP257 you should have a sentence here, hence it should end with a period:
```suggestion
    """Test that lenient_lowercase() proper results."""
```
Also, let's rephrase it to contain useful info.

**Review by webknjaz — DISMISSED** [review]
You have a positive test here. Are there any cases where it'd raise an exception? Those would need negative test cases.

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019, Andrew Klychkov @Andersson007 <aaklychkov@mail.ru>
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
```
Comment (samdoran) [inline test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]: Can you change the license to simplified GPL? The BSD license is only used in `module_utils`. The tests themselves should use the GPL 3.0 license.
```suggestion
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py (hunk 1 above)
Comment (samdoran) [reply test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]: 👍 to not needing this

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+INPUT_LIST = [
+    u'HELLO',
+    u'Ёлка',
+    u'cafÉ',
+    b'HELLO',
+    1,
+    {1: 'Dict'},
+    True,
+    [1],
```
Comment (samdoran) [inline test/units/module_utils/common/text/formatters/test_lenient_lowercase.py:24]: Add a float test case.
```suggestion
    [1],
    3.14159,
```

**Review by samdoran — CHANGES_REQUESTED** [review]
We should probably also add a function and data set for sending bad data to the function, something like `lenient_lowercase(1)` which raises `TypeError`.

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py (hunk 2 above)
Comment (Andersson007) [reply test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]: @webknjaz , IMO pytest.mark.parametrize is not suitable for this particular case, I can't imagine how to implement this.
The input list is passed once, then we just compare each element of it with a corresponding element of the expected list

**Comment by Andersson007** [comment]
@webknjaz @samdoran thank you for the suggestions about sending bad data to the function. Done

**Comment by Andersson007** [comment]
ready_for_review

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py (hunk 2 above)
Comment (webknjaz) [reply test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]: You mess up separate test cases into one. While they should show up separately. Also, one failure shouldn't affect other tests.

Comment (webknjaz) [reply test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]:
```suggestion
@pytest.mark.parametrize(
    'input_value,expected_outcome',
    (
        (i1, r1),
        (i2, r2),
       ...etc
    ),
)
def test_lenient_lowercase(input_value, expected_outcome):
```

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+def test_lenient_lowercase():
+    """Test that lenient_lowercase() proper results."""
+    output_list = lenient_lowercase(INPUT_LIST)
+    for out_elem, exp_elem in zip(output_list, EXPECTED_LIST):
+        assert out_elem == exp_elem
+
+
+@pytest.mark.parametrize('input_data', [1, False, 1.001, 1j, datetime.now(), ])
+def test_lenient_lowercase_unexpected_data_type(input_data):
+    """Test passing objects of unexpected types to lenient_lowercase()."""
```
Comment (webknjaz) [inline test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]: s/unexpected/illegal/

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    with pytest.raises(TypeError):
```
Comment (webknjaz) [inline test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]: Plz also use `match` arg here

File: test/units/module_utils/common/text/formatters/test_lenient_lowercase.py (hunk 2 above)
Comment (Andersson007) [reply test/units/module_utils/common/text/formatters/test_lenient_lowercase.py]: thank you, PTAL

**Comment by Andersson007** [comment]
ready_for_review

**Review by samdoran — APPROVED** [review]
